README.md
# Banner Time

<div>
<img src="https://sandywyper.github.io/Banner-Time/banner-time.jpg" style="max-width: 400px;width:100%">
</div>

**Version 1.0.0** - [Change log](CHANGELOG.md)

A simple sale baner that can be scheduled to show for a set time.

- [Example 1 - Fade](https://codepen.io/SandyWyper/full/yLLGzOx)
- [Example 2 - Slide](https://codepen.io/SandyWyper/full/qBBLaVg)

## Getting started

### Set-up

1. Download Banner-Time and put the '/dist/bannerTime.js' file in your site directory.
2. Reference this file from a script tag below where you include Jquery.

```
<script src="https://ajax.googleapis.com/ajax/libs/jquery/3.4.1/jquery.min.js"></script>
<script src="bannerTime.js"></script>
```

If you dont have JQuery in your project already, find it [here](https://jquery.com/).

3. Next you must initialise the banner. `const myBanner = new Banner()` This will create and insert the banner into your site using the default options. You should see the banner transition from 'Stop!!' to 'Banner Time!!'.

4. Insert your banner settings.

```
    const bannerSettings =  {
      startTime: [2019, 10, 9],
      endTime: [2020, 0, 1],
      CSS: {
        "border-bottom": "2px solid black",
        "background-color": "darkgrey",
        color: "white"
      },
      transition: {
        type: "slide",
        displayDuration: 5000,
        speed: 700,
        interval: 900
      }
    }
    const myBanner = new Banner(bannerSettings);
```

You can set up as many banners as you like, and the one with the most recent start time (and the 'endTime' has not yet expired) will display.

```
const firstBanner = {
  startTime: [2019,11,20],
   bannerText: {
        desktop: ["5 Days Until Xmas", "Shop now"],
        mobile: ["5 Days To Go!", "Xmas time!"]
      }
}
const secondBanner = {
  startTime: [2019,11,21],
   bannerText: {
        desktop: ["4 Days Until Xmas", "Shop now"],
        mobile: ["4 Days To Go!", "Xmas time!"]
      }
}
const thirdBanner = {
  startTime: [2019,11,22],
   bannerText: {
        desktop: ["3 Days Until Xmas", "Shop now"],
        mobile: ["3 Days To Go!", "Xmas time!"]
      }
}

const myXmasBanner = new Banner(firstBanner, secondBanner, thirdBanner);
```

5. If your shop isn't GTM then be sure to set your time zone in the settings `timeZone: -8`. This will adjust any start and end times accordingly.

6. Remeber to set the text output for both desktop and mobile screens.

---

### Settings

| Option           | Type                | Default                                                           | Description                                                                                                                                                                                                                                                                                                                                                                                                                                         |
| ---------------- | ------------------- | ----------------------------------------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| startTime        | number array        | [2019]                                                            | [**Individual date and time component values.**](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date#Individual_date_and_time_component_values "MDN Web Docs") Must be an array of numbers to represent the date and time. `[YYYY,MM,DD,hr,min]` (year, month, day, hour, minute, second, and millisecond) Any missing fields are given the lowest possible value (1 for the day and 0 for every other component. |
| endTime          | number array        | [2080]                                                            | The time the banner will finish displaying. See startTime for formatting.                                                                                                                                                                                                                                                                                                                                                                           |
| timeZone         | number              | 0                                                                 | Times are calculated in [UTC](http://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/UTC#description "MDN Web Docs"). Time zone of the shops(servers) location should be provided for corrrect start stop times. Number between -12 and 12.                                                                                                                                                                           |
| bannerLink       | string              |                                                                   | Enter the path of your desired link. Eg. `bannerLink: "/products/fidget-spinner"`                                                                                                                                                                                                                                                                                                                                                                   |
| **transition**   | object (nested..?)  | The default values of; **type, displayDuration, interval, speed** | A nested object with the keys and values for that transition. Declare the ones you need and the rest will remain default.                                                                                                                                                                                                                                                                                                                           |
| type             | string              | "fade"                                                            | The type of transition between each slide.                                                                                                                                                                                                                                                                                                                                                                                                          |
| display duration | number              | 5000                                                              | Time in milliseconds that each slide shows for before the next transition begins.                                                                                                                                                                                                                                                                                                                                                                   |
| interval         | number              | 800                                                               | Time in milliseconds between the slides transition out and in.                                                                                                                                                                                                                                                                                                                                                                                      |
| speed            | number              | 1000                                                              | Time in milliseconds of the transitions.                                                                                                                                                                                                                                                                                                                                                                                                            |
| **CSS**          | object (nested...?) | Any CSS can be added written as string key-value pairs; \*\*      | All the CSS you want attribute to each slide.                                                                                                                                                                                                                                                                                                                                                                                                       |
| any CSS property | string              |                                                                   |

---

#### Contributing

PLEASE review CONTRIBUTING.markdown prior to requesting a feature, filing a pull request or filing an issue.

---

#### Acknowledgements

This projected started out by following a modal window [tutorial](https://scotch.io/tutorials/building-your-own-javascript-modal-plugin) written by [Ken Wheeler](https://twitter.com/ken_wheeler). Then, in order to follow an example of an open source project, I have borrowed heavily from his [Slick Slider](https://github.com/kenwheeler/slick) repo. Big thanks to him and anyone else who would like to contribute.

## License and copyright

&copy; Alexander Wyper

Licensed under MIT [License](LICENSE).


CHANGELOG.md
## Change Log

#### Version: 1.0.0 - December 2019

- Initial release.

<!-- - development javascript file
- babel setup
- minified parsed js file
- README
- LICENSE
- CHANGELOG -->

src/bannerTime.js
(function() {
  "use:strict";
  // build stylesheet
  let styleBanner = document.createElement("style");
  styleBanner.innerHTML =
    "#banner-time{user-select: none;overflow:hidden;}#banner-time a{color: inherit;text-decoration: none;}#banner-time #desktop-banner-slides {display: none !important;}#banner-time #mobile-banner-slides {display: block !important;}@media only screen and (min-width: 768px) {#banner-time #desktop-banner-slides {display: block !important;}#banner-time #mobile-banner-slides{display: none !important;}}";
  // build banner element
  let bannerDiv = document.createElement("div");
  bannerDiv.id = "banner-time";
  // insert elements into the dom
  document.head.appendChild(styleBanner);
  document.body.prepend(bannerDiv);

  this.Banner = function() {
    // option defaults
    let defaults = {
      startTime: [2019],
      endTime: [2080],
      timeZone: 0,
      bannerLink: "",
      transition: {
        type: "fade",
        displayDuration: 5000,
        interval: 800,
        speed: 1000
      },
      bannerText: {
        desktop: ["STOP!!", "BANNER TIME!!"],
        mobile: ["STOP!!", "BANNER TIME!!"]
      },
      CSS: {
        display: "flex",
        "justify-content": "center",
        "align-items": "center",
        background: "orangeRed",
        color: "white",
        "white-space": "nowrap",
        height: "43px",
        "line-height": "30px",
        width: "100%",
        "font-family": "inherit",
        "font-size": "30px",
        "z-index": "25"
      }
    };

    // Ensure that the banner that is run is the correct one:
    // Eg. Most recent start time, and end time has not passed
    let relevantBannerIndex = getRelevantBannerIndex(arguments);
    // Create options by extending defaults with the passed in arugments
    if (
      arguments[relevantBannerIndex] &&
      typeof arguments[relevantBannerIndex] === "object"
    ) {
      this.options = extendDefaults(defaults, arguments[relevantBannerIndex]);
    } else {
      this.options = defaults;
    }

    // JQuery method to extend defaults with user options
    function extendDefaults(defaults, userOptions) {
      $.extend(true, defaults, userOptions);

      return defaults;
    }

    let timeWindow = checkTime(
      this.options.startTime,
      this.options.endTime,
      this.options.timeZone
    );

    if (timeWindow) {
      initBanner(this.options);
    } else {
      console.log("not showing promotion banner");
    }
  };
})();

function checkTime(starts, ends, timeZone) {
  //take date input from settings.
  //check for timezone option and adjust start/finsh dates to be UTC.
  const adjustTime = (time, timezone) => {
    let newTime = [];
    while (time.length < 4) {
      if (time.length === 2) {
        time.push(1);
      } else {
        time.push(0);
      }
    }

    time.forEach(function(int, index) {
      if (index === 3) {
        newTime.push(int + timezone);
      } else {
        newTime.push(int);
      }
    });

    return newTime;
  };

  let showBanner = new Date(Date.UTC(...adjustTime(starts, timeZone)));
  let hideBanner = new Date(Date.UTC(...adjustTime(ends, timeZone)));
  showBanner = showBanner.getTime();
  hideBanner = hideBanner.getTime();
  //when web user's page loads check banner start UTC time compared to current UTC and display banner or not.
  const timeNow = Date.now();
  if (timeNow > showBanner && timeNow < hideBanner) {
    return true;
  } else {
    return false;
  }
}

function getRelevantBannerIndex() {
  let mostRecentStartTime = 0;
  let relevantBannerIndex = 0;
  const timeNow = Date.now();

  Array.from(...arguments).forEach(function(arg, index) {
    let startTimeGiven = arg.startTime || [2019];
    startTimeGiven = new Date(Date.UTC(...startTimeGiven));
    startTimeGiven = startTimeGiven.getTime();
    let endTimeGiven = arg.endTime || [2080];
    endTimeGiven = new Date(Date.UTC(...endTimeGiven));
    endTimeGiven = endTimeGiven.getTime();

    if (mostRecentStartTime < startTimeGiven && timeNow < endTimeGiven) {
      mostRecentStartTime = startTimeGiven;
      relevantBannerIndex = index;
    }
  });

  return relevantBannerIndex;
}

function initBanner(options) {
  // Fill DOM element
  let $banner = $("#banner-time");

  let bannerHTML = buildBannerElements(
    options.bannerText.desktop,
    options.bannerText.mobile,
    options.bannerLink
  );

  $banner.append(bannerHTML);
  $banner.css(options.CSS);

  transition(options.transition, options.bannerText.desktop.length, "desktop");
  transition(options.transition, options.bannerText.mobile.length, "mobile");
}

function buildBannerElements(deskText, mobText, link) {
  let bannerConstructed = "";

  if (deskText) {
    bannerConstructed += '<div id="desktop-banner-slides">';

    deskText.forEach(function(text, index) {
      bannerConstructed += `<div class="js-slide-${index}" 
        ${index > 0 ? 'style="display:none"' : 'style="display:block"'}>`;
      if (link) {
        bannerConstructed += `<a href="${link}"><p>${text}</p></a></div>`;
      } else {
        bannerConstructed += `<p>${text}</p></div>`;
      }
    });

    bannerConstructed += `</div>`;
  }

  if (mobText) {
    bannerConstructed += `<div id="mobile-banner-slides">`;

    mobText.forEach(function(text, index) {
      bannerConstructed += `<div class="js-slide-${index}" ${
        index > 0 ? 'style="display:none"' : 'style="display:block"'
      }>`;
      if (link) {
        bannerConstructed += `<a href="${link}"><p>${text}</p></a></div>`;
      } else {
        bannerConstructed += `<p>${text}</p></div>`;
      }
    });

    bannerConstructed += `</div>`;
  }

  return bannerConstructed;
}

function transition(transition, numberOfSlides, screen) {
  if (numberOfSlides > 1) {
    switch (transition.type) {
      case "fade":
        fade(transition, numberOfSlides, screen);
        break;
      case "slide":
        slide(transition, numberOfSlides, screen);
      // *** more transition options to be added ***
    }
  }
}

function fade(transition, numberOfSlides, screen) {
  let count = 0;
  let $currentSlide;
  runTransition();

  function runTransition() {
    $currentSlide = $(`#${screen}-banner-slides .js-slide-${count}`);

    $currentSlide
      .delay(transition.interval)
      .fadeIn(transition.speed, function() {
        $currentSlide
          .delay(transition.displayDuration)
          .fadeOut(transition.speed, function() {
            count = count < numberOfSlides - 1 ? count + 1 : 0;
            runTransition();
          });
      });
  }
}

function slide(transition, numberOfSlides, screen) {
  let count = 0;
  let $currentSlide;
  $(`#${screen}-banner-slides`).css({
    position: "relative",
    width: "100%",
    height: "100%"
  });

  styleSlides(screen);

  runTransition();

  function runTransition() {
    $currentSlide = $(`#${screen}-banner-slides .js-slide-${count}`);
    $currentSlide
      .animate(
        {
          left: $currentSlide.parent().width() / 2 - $currentSlide.width() / 2
        },
        transition.speed
      )
      .delay(transition.displayDuration)
      .animate({ left: "300%" }, transition.speed, function() {
        setTimeout(function() {
          $currentSlide.css({ left: "-200%" });
          count = count < numberOfSlides - 1 ? count + 1 : 0;
          $(this).stop();
          runTransition();
        }, transition.interval);
      });
  }
}

function styleSlides(screen) {
  $(`#${screen}-banner-slides > div`).css({
    position: "absolute",
    height: "100%",
    left: "-200%",
    display: "flex",
    "align-items": "center",
    "justify-content": "center"
  });
}
